kboot status report for 2023 Q2

bsdimp opened the pull request:
oh, and a little before too, but this is my first one

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+Contact: Warner Losh <imp@bsdimp.com>
+
+Links: +
+link:https://www.linuxboot.org/[LinuxBoot Project] +
```
```suggestion
link:https://www.linuxboot.org/[LinuxBoot Project] URL: link:https://www.linuxboot.org/[]+
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+
+Links: +
+link:https://www.linuxboot.org/[LinuxBoot Project] +
+link:https://docs.google.com/presentation/d/1N5Jp6XzYWv9Z9RhhETC-e6tFkqRHvp-ldRDW_9h2JCw/edit?usp=sharing[BSDCan 2024 kboot talk slides]
```
```suggestion
link:https://docs.google.com/presentation/d/1N5Jp6XzYWv9Z9RhhETC-e6tFkqRHvp-ldRDW_9h2JCw/edit?usp=sharing[BSDCan 2024 kboot talk slides] URL: link:https://docs.google.com/presentation/d/1N5Jp6XzYWv9Z9RhhETC-e6tFkqRHvp-ldRDW_9h2JCw/edit?usp=sharing[]
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+link:https://docs.google.com/presentation/d/1N5Jp6XzYWv9Z9RhhETC-e6tFkqRHvp-ldRDW_9h2JCw/edit?usp=sharing[BSDCan 2024 kboot talk slides]
+
+LinuxBoot is an effort to create a clean, robust, auditable and repeatable boot firmware.
+What originally started as a specific project at Google has grown to encompas any boot environment that uses Linux to launch the final operating system.
```
```suggestion
What originally started as a specific project at Google has grown to encompass any boot environment that uses Linux(R) to launch the final operating system.
```

bsdimp replied:
Not going to add (r) to Linux. It's not needed and I'm under no obligation to do so.

grahamperrin replied:
Agreed. I wasn't sure, IIRC status reports are quite easygoing in this respect.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+
+LinuxBoot is an effort to create a clean, robust, auditable and repeatable boot firmware.
+What originally started as a specific project at Google has grown to encompas any boot environment that uses Linux to launch the final operating system.
+This sort of environment is available on many platforms, and in some cases it he only available boot environment.
```
```suggestion
This sort of environment is available on many platforms, and in some cases is the only available boot environment.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+LinuxBoot is an effort to create a clean, robust, auditable and repeatable boot firmware.
+What originally started as a specific project at Google has grown to encompas any boot environment that uses Linux to launch the final operating system.
+This sort of environment is available on many platforms, and in some cases it he only available boot environment.
+In addition, some embedded boxes have a linux boot environment hard-coded that's quite hard to change, and being able to reboot into FreeBSD is desirable.
```
```suggestion
In addition, some embedded boxes have a Linux boot environment hard-coded that is quite hard to change, and being able to reboot into FreeBSD is desirable.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+This sort of environment is available on many platforms, and in some cases it he only available boot environment.
+In addition, some embedded boxes have a linux boot environment hard-coded that's quite hard to change, and being able to reboot into FreeBSD is desirable.
+
+The old Sony Playstation 3 port used a boot loader called 'kboot' to accomplish this in that environment.
```
```suggestion
The old Sony PlayStation(R) 3 port used a boot loader called 'kboot' to accomplish this in that environment.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+The old Sony Playstation 3 port used a boot loader called 'kboot' to accomplish this in that environment.
+That code has been greatly expanded, made generic with easily replaceable per-architecture plug ins.
+The normal FreeBSD /boot/loader is built as a Linux binary that reads in the FreeBSD kernel, modules and tunables.
+It places them into memory as if it were running in a pre-boot environment and then loads that image into the Linux kernel with kexec_load(2) and does a special reboot to that kernel image.
```
```suggestion
It places them into memory as if it were running in a pre-boot environment and then loads that image into the Linux kernel with man:kexec_load[2] and does a special reboot to that kernel image.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+It places them into memory as if it were running in a pre-boot environment and then loads that image into the Linux kernel with kexec_load(2) and does a special reboot to that kernel image.
+For UEFI enabled systems, it passes the UEFI memory table, and pointer to UEFI runtime services to the new kernel.
+
+It supports loading files from the host's filesystem, from any loader(8) supported filesystem on the host's block devices (including ZPOOLS that span multiple devices), from ram disk images and from files downloaded over the network.
```
```suggestion
It supports loading files from the host's filesystem, from any man:loader[8]-supported filesystem on the host's block devices (including pools that span multiple devices), from ram disk images and from files downloaded over the network.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+It supports loading files from the host's filesystem, from any loader(8) supported filesystem on the host's block devices (including ZPOOLS that span multiple devices), from ram disk images and from files downloaded over the network.
+Any mix of these is available, so configuraiton overrides can be loaded from the host's filesystem, but the kernel can be loaded from dedicated storage (say NVME) or a ram disk image.
+It supports a host console running over stdin/stdout.
+It supports explicit locations like `/dev/nvme0ns1:/boot/loader/gerbil.conf` to be explicit about where to load filesystems from.
```
```suggestion
It supports explicit locations such as `/dev/nvme0ns1:/boot/loader/gerbil.conf` for where to load filesystems from.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+Any mix of these is available, so configuraiton overrides can be loaded from the host's filesystem, but the kernel can be loaded from dedicated storage (say NVME) or a ram disk image.
+It supports a host console running over stdin/stdout.
+It supports explicit locations like `/dev/nvme0ns1:/boot/loader/gerbil.conf` to be explicit about where to load filesystems from.
+If supports ZFS boot environments, including the boot-once feature.
```
```suggestion
It supports ZFS boot environments, including the boot-once feature.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+It supports explicit locations like `/dev/nvme0ns1:/boot/loader/gerbil.conf` to be explicit about where to load filesystems from.
+If supports ZFS boot environments, including the boot-once feature.
+
+FreeBSD/aarch64 now can boot from Linux in a LinuxBoot environment, with comparable support and functionality as loader.efi.
```
```suggestion
FreeBSD/aarch64 now can boot from Linux in a LinuxBoot environment, with support and functionality comparable to man:loader.efi[8].
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+FreeBSD/aarch64 now can boot from Linux in a LinuxBoot environment, with comparable support and functionality as loader.efi.
+Memory layout passed in for GICv3 workarounds.
+Need patch for aarch64 kernel for the GICv3 workaround.
+link:https://reviews.freebsd.org/D40902
```
```suggestion
link:https://reviews.freebsd.org/D40902[]
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+The amd64 boot environment has more quirks and details to get right than the aarch64 environment given its much older lineage.
+The kernel boots part way before panicing because things aren't setup correctly.
+
+powerpc proudly builds, but nothing more of its state is known.
```
```suggestion
PowerPC(TM) proudly builds, but nothing more of its state is known.
```

bsdimp replied:
Same answer: not going to add tm here. It's not needed.

grahamperrin replied:
™ removed from suggestion.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+Unifying the kboot eand efi copies is needed. +
+This is complicated because they aren't identical.
+ 1. It would be nice to add riscv64 support
+ 1. powerpc testing (its been untested since the refactoring started).
```
```suggestion
 1. PowerPC testing (it has been untested since the refactoring started).
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+This is complicated because they aren't identical.
+ 1. It would be nice to add riscv64 support
+ 1. powerpc testing (its been untested since the refactoring started).
+ 1. Help creating a script to repackage EDK-II image (say from QEMU) as a linux-boot image with a linux kernel built on FreeBSD for CI testing
```
```suggestion
 1. Help creating a script to repackage EDK-II image (say from QEMU) as a linux-boot image with a Linux kernel built on FreeBSD for CI testing.
```

grahamperrin replied:
Here, should _linux-boot_ be _LinuxBoot_?

grahamperrin reviewed (changes requested):
Markup and oddments.

bsdimp commented:
Since I'm not selling anything, I don't need r or tm. I don't want to imply I'm selling anything by using them. The other comments look good and I'll likely include them all...

grahamperrin commented:
Sorry: I hid my comment at <https://github.com/freebsd/freebsd-doc/pull/202#pullrequestreview-1520414964> not realising that doing so hides everything beneath (including non-resolved stuff) … can't find a way to un-hide.

lsalvadore commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+ 1. Currently, there is no documentation on how to use loader.kboot, nor on how to create test images.
+ 1. Finish amd64 support.
+ 1. The current elf arch-specific metadata code is copied from efi. +
+Unifying the kboot eand efi copies is needed. +
```
s/eand/and/

grahamperrin replied:
```suggestion
Unifying the kboot and efi copies is needed. +
```

- resolve https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1257443258

bsdimp replied:
done

lsalvadore commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+ 1. Finish amd64 support.
+ 1. The current elf arch-specific metadata code is copied from efi. +
+Unifying the kboot eand efi copies is needed. +
+This is complicated because they aren't identical.
```
Please do not use contractions:
s/aren't/are not/

grahamperrin replied:
```suggestion
This is complicated because they are non-identical.
```

- resolve https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1257442298

bsdimp replied:
reworded as
```
 1. The current elf arch-specific metadata code is copied from efi. +
Unifying the kboot and efi copies is needed. +
While they are mostly the same, compile-time differences remain that	complicate sharing. +
In addtion, the build infrastructure makes sharing awkward.
```

lsalvadore reviewed (changes requested):
Thanks for your report. I made a few suggestions too and I have also spotted a few typos.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+It supports loading files from the host's filesystem, from any loader(8) supported filesystem on the host's block devices (including ZPOOLS that span multiple devices), from ram disk images and from files downloaded over the network.
+Any mix of these is available, so configuraiton overrides can be loaded from the host's filesystem, but the kernel can be loaded from dedicated storage (say NVME) or a ram disk image.
+It supports a host console running over stdin/stdout.
+It supports explicit locations like `/dev/nvme0ns1:/boot/loader/gerbil.conf` to be explicit about where to load filesystems from.
```
If I understand correctly: it's a method of specifying a location, in the context of kboot, however it's _not_ a conventional path to a file. 

----

As the first command below gets me nowhere, so neither `file /dev/nvme0ns1:/boot/loader/gerbil.conf` nor (more simply) `file /dev/nvme0ns1:/boot/loader` can get anyone anywhere. 

```text
% file /dev/ada0p3.eli:/boot/loader
/dev/ada0p3.eli:/boot/loader: cannot open `/dev/ada0p3.eli:/boot/loader' (No such file or directory)
% file /boot/loader
/boot/loader: FreeBSD/i386 demand paged executable
% lsblk ada0
DEVICE         MAJ:MIN SIZE TYPE                                          LABEL MOUNT
ada0             0:129 932G GPT                                               - -
  ada0p1         0:131 260M efi                                    gpt/efiboot0 -
  <FREE>         -:-   1.0M -                                                 - -
  ada0p2         0:133  16G freebsd-swap                              gpt/swap0 SWAP
  ada0p2.eli     0:167  16G freebsd-swap                                      - SWAP
  ada0p3         0:135 915G freebsd-zfs                                gpt/zfs0 <ZFS>
  ada0p3.eli     0:142 915G -                                                 - -
  <FREE>         -:-   708K -                                                 - -
% 
```

HTH

bsdimp replied:
close enough to real filename

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+
+FreBSD/amd64 in progress, maybe 80% done.
+The amd64 boot environment has more quirks and details to get right than the aarch64 environment given its much older lineage.
+The kernel boots part way before panicing because things aren't setup correctly.
```
```suggestion
The kernel boots part way before a panic because things aren't setup correctly.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+It supports explicit locations like `/dev/nvme0ns1:/boot/loader/gerbil.conf` to be explicit about where to load filesystems from.
+If supports ZFS boot environments, including the boot-once feature.
+
+FreeBSD/aarch64 now can boot from Linux in a LinuxBoot environment, with comparable support and functionality as loader.efi.
```
- https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1256761248

bsdimp replied:
went with @grahamperrin 's suggestion instead that he linked.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+
+Help needed:
+
+ 1. Currently, there is no documentation on how to use loader.kboot, nor on how to create test images.
```
If I understand correctly: 

- this is not a filename
- it may become a manual page, e.g. loader.kboot(8) | `man:loader.kboot[8]`

– assuming section 8 of the manual (compare with [loader.efi(8)](https://man.freebsd.org/cgi/man.cgi?query=loader.efi&sektion=8&manpath=freebsd-release)).

bsdimp replied:
I've updated this with
```
 1. man:loader.kboot[8]	needs to be written. It	should document	how to use loader.kboot, how to	create images, and the different use cases that	work today.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+
+FreBSD/amd64 in progress, maybe 80% done.
+The amd64 boot environment has more quirks and details to get right than the aarch64 environment given its much older lineage.
+The kernel boots part way before panicing because things aren't setup correctly.
```
```suggestion
The kernel boots partially then panics, because setup is incorrect.
```

- resolve <https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1257443036>
- avoid the _aren't_ contraction of _are not_
- avoid the word _things_ – the suggestion above might be preferable to the alternative below.

> The kernel boots partially then panics, because things are setup incorrectly.

----

(My first thought was to replace _panicing_, which LanguageTool does detect as a possible spelling mistake, with _a panic_ … a few hours later, thoughts changed.)

bsdimp replied:
```
The kernel panics partway through initialization since	all the	prerequisites from the boot loader have	not been discovered and	implemented.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+
+The old Sony Playstation 3 port used a boot loader called 'kboot' to accomplish this in that environment.
+That code has been greatly expanded, made generic with easily replaceable per-architecture plug ins.
+The normal FreeBSD /boot/loader is built as a Linux binary that reads in the FreeBSD kernel, modules and tunables.
```
```suggestion
The normal FreeBSD [.filename]#/boot/loader# is built as a Linux binary that reads in the FreeBSD kernel, modules and tunables.
```

- resolve https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1257442111

grahamperrin replied:
Cross-reference <https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=266881#c3> re: discovery of markup such as this.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,48 @@
+For UEFI enabled systems, it passes the UEFI memory table, and pointer to UEFI runtime services to the new kernel.
+
+It supports loading files from the host's filesystem, from any loader(8) supported filesystem on the host's block devices (including ZPOOLS that span multiple devices), from ram disk images and from files downloaded over the network.
+Any mix of these is available, so configuraiton overrides can be loaded from the host's filesystem, but the kernel can be loaded from dedicated storage (say NVME) or a ram disk image.
```
```suggestion
Any mix of these is available, so configuration overrides can be loaded from the host's filesystem, but the kernel can be loaded from dedicated storage (say NVME) or a ram disk image.
```

- resolve https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1257442298

grahamperrin reviewed (changes requested):
Thanks to @lsalvadore for the catches.

bsdimp commented:
pushed update that I claim addresses all the feedback here.
Please let me know if I missed anything (I tried to include it all, except trademarks, etc since any such trademark used is done in a descriptive way referring to a family of products, not one specific one, in ways that would be hard to construe as endorsement).
I've also tweaked some overly verbose and passive language

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+link:https://reviews.freebsd.org/D40902[]
+
+FreBSD/amd64 support is in progress and is maybe 80% done.
+The amd64 boot environment places more requirements on the boot loader to provide data for the kenrel than aarch64 due to amd64 being an older port.
```
```suggestion
The amd64 boot environment places more requirements on the boot loader to provide data for the kernel than aarch64, due to amd64 being an older port.
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+ 1. Finish amd64 support.
+ 1. The current elf arch-specific metadata code is copied from efi. +
+Unifying the kboot and efi copies is needed. +
+While they are mostly the same, compile-time differences remain that complicate sharing. +
```
```suggestion
While they are mostly the same, sharing is complicated by remaining compile-time differences. +
```

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+The amd64 boot environment places more requirements on the boot loader to provide data for the kenrel than aarch64 due to amd64 being an older port.
+All sources for data in the BIOS enviornment had to be provided by the boot loader since the kernel had no access to them from long mode.
+While UEFI and ACPI provide ways for the kernel to get this data, much of the data must still be provided by the boot loader.
+The kernel panics during initialization since all these prerequisites have not been discovered and implemented.
```
```suggestion
As not all prerequisites have been discovered and implemented, the kernel panics during initialization.
```

bsdimp replied:
How about
```
Loader.kboot does not yet implement the entire amd64 boot protocol, so the kernel still panics.
```

grahamperrin replied:
Suggestion revised, NB: 

- lowercase `l`
- `[.filename]#⋯#` markup seems reasonable here.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+In addtion, the build infrastructure makes sharing awkward.
+ 1. It would be nice to add riscv64 support
+ 1. PowerPC testing (it has been untested since the refactoring started).
+ 1. Help creating a script to repackage EDK-II image (say from QEMU) as a linux-boot image with a Linux kernel built on FreeBSD for CI testing.
+ 1. Help testing it from the coreboot LinuxBoot
```
```suggestion
 1. Creating a script to repackage EDK-II image (say, from QEMU) as a linux-boot image with a Linux kernel built on FreeBSD for CI testing.
 1. Testing it from the coreboot LinuxBoot.
```

- As the list is preceded by 'Help needed:', so list items need not begin with 'Help …'.

grahamperrin reviewed (changes requested):
Thanks,

<https://github.com/freebsd/freebsd-doc/pull/202#issuecomment-1628329016>

> … Please let me know if I missed anything …

One typo (kenrel), so I took the opportunity to speed-read the whole thing with fresh eyes. 

Take only what you want from the suggested changes, no need to explain anyth discards (unless you want to).

> … if all my 'accept' generated insane spam)

It's a neat feature, I never thought of it as spammy :-)

bsdimp commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+It places them into memory as if it were running in a pre-boot environment and then loads that image into the Linux kernel with man:kexec_load[2] and does a special reboot to that kernel image.
+For UEFI enabled systems, it passes the UEFI memory table, and pointer to UEFI runtime services to the new kernel.
+
+It supports loading files from the host's filesystem, from any man:loader[8]-supported filesystem on the host's block devices (including pools that span multiple devices), from ram disk images and from files downloaded over the network.
```
I like your change. but would say 'from the network' rather than 'from networked files'. Comments?

grahamperrin replied:
Thanks. Suggestion revised. Have I twisted the meaning?

bsdimp commented:
Accepted all but two. What do you think of my suggested changes to your suggested change ... If you like them, just make them a suggestion and I'll accept it. If not, we'll chat :)

pauamma commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+
+LinuxBoot is an effort to create a clean, robust, auditable and repeatable boot firmware.
+What originally started as a specific project at Google has grown to encompass any boot environment that uses Linux to launch the final operating system.
+Many platfors now support this environment, in some cases it is the only available boot environment.
```
```suggestion
Many platforms now support this environment, and in some cases it is the only available boot environment.
```

pauamma commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+That code has been greatly expanded, made generic with easily replaceable per-architecture plug ins.
+The normal FreeBSD [.filename]#/boot/loader# is built as a Linux binary that reads in the FreeBSD kernel, modules and tunables.
+It places them into memory as if it were running in a pre-boot environment, then loads that image into the Linux kernel with man:kexec_load[2] and does a special reboot to that image.
+For UEFI enabled systems, it passes the UEFI memory table, and pointer to UEFI runtime services to the new kernel.
```
```suggestion
For UEFI-enabled systems, it passes the UEFI memory table and pointer to UEFI runtime services to the new kernel.
```

grahamperrin replied:
If we add a second comma, then it probably needs a third. 

Alternatively: 

> With UEFI-enabled systems, the UEFI memory table and a pointer to UEFI runtime services are passed to the new kernel.

pauamma reviewed (commented):
Spelling and punctuation nits.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+
+The old Sony PlayStation 3 port used a boot loader called 'kboot' to boot the FreeBSD port from its Linux kernel (all predating the LinuxBoot project).
+That code has been greatly expanded, made generic with easily replaceable per-architecture plug ins.
+The normal FreeBSD [.filename]#/boot/loader# is built as a Linux binary that reads in the FreeBSD kernel, modules and tunables.
+It places them into memory as if it were running in a pre-boot environment, then loads that image into the Linux kernel with man:kexec_load[2] and does a special reboot to that image.
```
```suggestion

The normal FreeBSD [.filename]#/boot/loader# is built as a Linux binary that reads in the FreeBSD kernel, modules, and tunables.

LinuxBoot places the FreeBSD kernel, modules, and tunables into memory as if it were running in a pre-boot environment, then loads that image into the Linux kernel with man:kexec_load[2] and performs a special reboot to the image.
```

Disambiguate the words 'It' and 'them' … correct, or have I misunderstood?

(I might be slaving over this unnecessarily; the report acknowledges the need for a manual page.)

bsdimp replied:
as if loader.kboot were running as loader.efi and trying to load the kernel, tuneables and modules from there (similar memory layouts happen). I struggled with a way to say loader.kboot will present the same environment to the kernel's startup code after loader.kboot does its kexec that loader.efi does after loading everything and jumping to the kernel's startup code.

grahamperrin replied:
> … I struggled with a way to say …

How about … a freehand sketch? (As if you have nothing better to do.) 

It needn't be perfect; the absence of loader.kboot(8) is uppermost amongst your pleas for help.

File an image somewhere under `/home/imp/public_html` then it can be linked from within the report. Just a thought.

lsalvadore replied:
If an image is to be included, please remember that the report will also be sent by email. So either we include a `link` reference or maybe a picture in Ascii art.

But a good wording is probably better.

lsalvadore commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+
+Help needed:
+
+ 1. man:loader.kboot[8] needs to be written. It should document how to use [.filename]#loader.kboot#, how to create images, and the use cases that work today.
```
```suggestion
 1. man:loader.kboot[8] needs to be written.
 It should document how to use [.filename]#loader.kboot#, how to create images, and the use cases that work today.
```
One sentence per line please, even in list elements.

grahamperrin replied:
Rolled into (superseded by) https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1271623265

lsalvadore commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+Unifying the kboot and efi copies is needed. +
+While they are mostly the same, sharing is complicated by remaining compile-time differences. +
+In addtion, the build infrastructure makes sharing awkward.
+ 1. It would be nice to add riscv64 support
```
```suggestion
 1. It would be nice to add riscv64 support.
```

grahamperrin replied:
Rolled into (superseded by) https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1271623265

lsalvadore reviewed (approved):
I have found two last minute small issues.
I think we can finally merge this report, is anything still missing?

grahamperrin commented:
> is anything still missing?

At a glance, at least: 

1. substantially, <https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1258915778>, which, as revised, is Warner's suggestion
2. nelgigibly, where <https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1264653383> begins to pay attention to an ordered list, now I can see a leading white space at each number.

----

<https://docs.asciidoctor.org/asciidoc/latest/syntax-quick-reference/#lists> example 31 presents what's below the lead for a simple ordered list: 

`. `

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+
+ 1. man:loader.kboot[8] needs to be written. It should document how to use [.filename]#loader.kboot#, how to create images, and the use cases that work today.
+ 1. Finish amd64 support.
+ 1. The current elf arch-specific metadata code is copied from efi. +
+Unifying the kboot and efi copies is needed. +
+While they are mostly the same, sharing is complicated by remaining compile-time differences. +
+In addtion, the build infrastructure makes sharing awkward.
```
Is the hardness `+` of line breaks intentional? 

Previewed: 

![image](https://github.com/freebsd/freebsd-doc/assets/192271/6c3e78f1-5d60-419b-851f-c638a86f9ad4)

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+PowerPC builds, but nothing more of its state is known.
+Attempts to acquire a suitable Playstation 3 proved to be too time consuming for the author.
+
+Help needed:
```
```suggestion
==== Help Needed
```

- a simple subheading
- and then, what's currently an ordered list of numbered items can become an unnumbered series of simple paragraphs. 

----

The final paragraph, sponsorship already in place, will not be misinterpreted as something with which you need help.

lsalvadore commented:
Thanks @bsdimp, @grahamperrin and @pauamma for your efforts to put in place a high quality report.

Do you think you can complete the last few things you need in the next few days? It is already late July, I would like to publish the status report within August 15th, or even better within the end of the month. It does not matter if it is not perfect, as long as the content is there.

If you do not have time and you want me to make the merge and fix any last minute thing, please just ask.

bsdimp commented:
I thought that Iwas done.  what's left?

bsdimp commented:
oh,  I  see. The last few suggestions are fine: the hard breaks aren't intentional and the new section is fine. I don't have time to revise and repush, thought.

grahamperrin commented on website/content/en/status/report-2023-04-2023-06/kboot.adoc:
```diff
@@ -0,0 +1,55 @@
+
+Help needed:
+
+ 1. man:loader.kboot[8] needs to be written. It should document how to use [.filename]#loader.kboot#, how to create images, and the use cases that work today.
+ 1. Finish amd64 support.
+ 1. The current elf arch-specific metadata code is copied from efi. +
+Unifying the kboot and efi copies is needed. +
+While they are mostly the same, sharing is complicated by remaining compile-time differences. +
+In addtion, the build infrastructure makes sharing awkward.
+ 1. It would be nice to add riscv64 support
+ 1. PowerPC testing (it has been untested since the refactoring started).
+ 1. Creating a script to repackage EDK-II image (say, from QEMU) as a linux-boot image with a Linux kernel built on FreeBSD for CI testing.
+ 1. Testing it from the coreboot LinuxBoot.
```
```suggestion

man:loader.kboot[8] needs to be written.
It should document how to use [.filename]#loader.kboot#, how to create images, and the use cases that work today.

Finish amd64 support.

The current elf arch-specific metadata code is copied from efi.
Unifying the kboot and efi copies is needed.
While they are mostly the same, sharing is complicated by remaining compile-time differences.
In addtion, the build infrastructure makes sharing awkward.

It would be nice to add riscv64 support.

PowerPC testing (it has been untested since the refactoring started).

Creating a script to repackage EDK-II image (say, from QEMU) as a linux-boot image with a Linux kernel built on FreeBSD for CI testing.

Testing it from the coreboot LinuxBoot.
```

- consistent with Warner's https://github.com/freebsd/freebsd-doc/pull/202#issuecomment-1646953449
- consistent with (complementary to) my https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1264692623
- consistent with Lorenzo's per-line markup from https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1264653383
- includes the additional full stop from Lorenzo's https://github.com/freebsd/freebsd-doc/pull/202#discussion_r1264653406
- seven paragraphs instead of points 1–7.

grahamperrin reviewed (changes requested):
Re: <https://github.com/freebsd/freebsd-doc/pull/202#issuecomment-1646953449>

> … repush …

If it simplifies things: 

1. the two-click routine for each acceptable suggestion, then
2. <https://patch-diff.githubusercontent.com/raw/freebsd/freebsd-doc/pull/202.diff> should be ready for Lorenzo to push (if everything that's required existed as a _suggested change_ for step 1).

grahamperrin commented:
My recent question to `FreeBSD Git administrators <gitadm@freebsd.org>` (addressee invisible at <https://lists.freebsd.org/archives/freebsd-doc/2023-July/004013.html>) was partly for people to learn whether a _maintainer_ can perform actions such as clicking to accept what's acceptable.

> GitHub: allow edits by maintainers

sergio-carlavilla commented:
Merged.
